db/src/metadata.rs

```diff
@@ -0,0 +1,379 @@
+//!
+//! This module recognizes, validates, applies, and reports on these mutations.
+
+use std::collections::{BTreeMap, BTreeSet};
+use std::collections::btree_map::Entry;
+
+use itertools::Itertools; // For join().
+
+use db;
+use entids;
+use errors::{
+    ErrorKind,
+    Result,
+};
+use mentat_core::{
+    Entid,
+    Schema,
+    TypedValue,
+    ValueType,
+};
+use schema::{
+    AttributeBuilder,
+    SchemaBuilding,
+};
+
+/// An alteration to an attribute.
+#[derive(Clone, Debug, Eq, Hash, Ord, PartialOrd, PartialEq)]
+pub enum AttributeAlteration {
+    /// From http://blog.datomic.com/2014/01/schema-alteration.html:
+    /// - rename attributes
+    /// - rename your own programmatic identities (uses of :db/ident)
+    /// - add or remove indexes
+    Index,
+    /// - add or remove uniqueness constraints
+    UniqueValue,
+    UniqueIdentity,
+    /// - change attribute cardinality
+    Cardinality,
+    /// - change whether history is retained for an attribute
+    NoHistory,
+    /// - change whether an attribute is treated as a component
+    IsComponent,
+}
+
+/// Summarizes changes to metadata such as a a `Schema` and (in the future) a `PartitionMap`.
+#[derive(Clone, Debug, Eq, Hash, Ord, PartialOrd, PartialEq)]
+pub struct MetadataReport {
+    /// Entids that have been flagged with [:db.part/db :db.install/attribute entid].
+    pub flagged_install: BTreeSet<Entid>,
+
+    // Entids that have been flagged with [:db.part/db :db.alter/attribute entid].
+    pub flagged_alter: BTreeSet<Entid>,
+
+    // Entids that were not present in the original `SchemaMap` that was mutated.
+    pub installed: BTreeSet<Entid>,
+
+    // Entids that were present in the original `SchemaMap` that was mutated, together with a
+    // representation of the mutations that were applied.
+    pub altered: BTreeMap<Entid, Vec<AttributeAlteration>>,
+}
+
+impl MetadataReport {
+    pub fn is_empty(&self) -> bool {
+        self.installed.is_empty() && self.altered.is_empty()
+    }
+}
+
+lazy_static! {
+    pub static ref ATTRIBUTE_SET: String = {
+        // let mut stmt = self.prepare_cached("SELECT e, a, v, value_type_tag, added FROM transactions WHERE tx = ? AND a IN 
+        format!("({}, {}, {}, {}, {}, {}, {}, {}, {}, {}, {}, {})",
+                entids::DB_ALTER_ATTRIBUTE,
+                entids::DB_CARDINALITY,
+                entids::DB_DOC,
+                entids::DB_FULLTEXT,
+                entids::DB_IDENT,
+                entids::DB_INDEX,
+                entids::DB_INSTALL_ATTRIBUTE,
+                entids::DB_INSTALL_PARTITION,
+                entids::DB_INSTALL_VALUE_TYPE,
+                entids::DB_IS_COMPONENT,
+                entids::DB_UNIQUE,
+                entids::DB_VALUE_TYPE)
+    };
+}
+
+/// Return `false` if the given reduced entity will not change the metadata: recognized idents,
+/// schema, partitions in the partition map.
+pub fn might_update_metadata(reduced: &db::ReducedEntity) -> bool {
```

Comment:
I'd suggest marking this as `#[inline]`, see if it speeds up the transactor tight loop.